pkg/cluster/advertise.go

```diff
@@ -3,46 +3,47 @@ package cluster
 import (
 	"net"
 
+	"strconv"
+
 	"github.com/hashicorp/go-sockaddr"
 	"github.com/pkg/errors"
 )
 
-// calculateAdvertiseAddress attempts to clone logic from deep within memberlist
-// (NetTransport.FinalAdvertiseAddr) in order to surface its conclusions to the
-// application, so we can provide more actionable error messages if the user has
-// inadvertantly misconfigured their cluster.
-//
-// https://github.com/hashicorp/memberlist/blob/022f081/net_transport.go#L126
-func calculateAdvertiseAddress(bindAddr, advertiseAddr string) (net.IP, error) {
+// CalculateAdvertiseAddress deduce the external, advertise address that should be routable from other components.
+func CalculateAdvertiseAddress(bindAddr, advertiseAddr string) (string, int, error) {
 	if advertiseAddr != "" {
-		ip := net.ParseIP(advertiseAddr)
-		if ip == nil {
-			return nil, errors.Errorf("failed to parse advertise addr '%s'", advertiseAddr)
+		advHost, advPort, err := net.SplitHostPort(advertiseAddr)
+		if err != nil {
+			return "", 0, errors.Wrap(err, "invalid advertise address")
 		}
-		if ip4 := ip.To4(); ip4 != nil {
-			ip = ip4
+		advIntPort, err := strconv.Atoi(advPort)
+		if err != nil {
+			return "", 0, errors.Wrapf(err, "invalid advertise address '%s', wrong port", advertiseAddr)
 		}
-		return ip, nil
+		return advHost, advIntPort, nil
+	}
+
+	// We expect bindAddr to be in form of ip:port.
+	bindHost, bindPort, err := net.SplitHostPort(bindAddr)
```

Comment:
Nit: check that port number is not `0`. This is sometimes used to signal to get any available port.